samza-api/src/main/java/org/apache/samza/operators/windows/BaseWindowFunction.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.windows;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.data.MessageEnvelope;
+
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ *  The base class of all types of {@link WindowFunction}s. Sub-classes can specify the default triggering semantics
+ *  for the {@link WindowFunction}, semantics for emission of early or late results and whether to accumulate or discard
+ *  previous results.
+ */
+@InterfaceStability.Unstable
+public class BaseWindowFunction<M extends MessageEnvelope, K, WV> implements WindowFunction<M, K, WindowKey<K>, WV, WindowOutput<WindowKey<K>, WV>> {
+
+  /**
+   * Defines the default triggering semantics for the {@link WindowFunction}.
+   */
+  private final List<TriggersBuilder.Trigger> defaultTriggers;
+
+
+  /**
+   * Defines the triggering semantics for emission of early or late results.
+   */
+  private List<TriggersBuilder.Trigger> earlyTriggers;
+  private List<TriggersBuilder.Trigger> lateTriggers;
+
+  /**
+   * Defines the fold function that is applied each time a {@link MessageEnvelope} is added to this window.
+   */
+  private final BiFunction<M, WV, WV> aggregator;
```

Comment:
> Is the intention to include cases that we are just accumulating the input values in a window w/o aggregate them into a single value? 
Actually the opposite - "Aggregate" implies accumulation to me, which is the opposite of what we intend here. I think we're on the same page about the intention for this method, just that the name means different things to us.

> I ended up picking foldFunction in the revised draft.
I think reduce can be implemented as a fold as well, so agree that foldFn is better than the overly general processor. If @nickpan47 agrees, maybe rename it everywhere consistently (e.g. in Windows) and document the semantics.